waiter/src/waiter/auth/composite.clj

```diff
@@ -0,0 +1,72 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.auth.composite
+  (:require [plumbing.core :as pc]
+            [waiter.auth.authentication :as auth]
+            [waiter.util.utils :as utils]))
+
+(defn- request->authentication
+  "Gets authentication service parameter for a request."
+  [request]
+  (get-in request [:waiter-discovery :service-parameter-template "authentication"]))
+
+(defrecord CompositeAuthenticator [authenticators default-authenticator]
+  auth/Authenticator
+
+  (process-callback [_ request scheme operation]
+    (if-let [authenticator (get authenticators (keyword scheme))]
+      (auth/process-callback authenticator request scheme operation)
+      (throw (ex-info (str "Unknown authentication scheme " scheme)
+                      {:composite-authenticators (keys authenticators)
+                       :scheme scheme
+                       :status 400}))))
+  (wrap-auth-handler [_ request-handler]
+    (let [default-handler (auth/wrap-auth-handler default-authenticator request-handler)
+          handlers (pc/map-vals #(auth/wrap-auth-handler % request-handler) authenticators)]
+      (fn composite-authenticator-handler [request]
+        (let [authentication (get-in request [:waiter-discovery :service-parameter-template "authentication"])]
```

Comment:
We can use `request->authentication` here.

Reply:
outdated